### PR TITLE
eof: Pass `_eofVersion` down to `libevmasm/Assembly` class

### PR DESCRIPTION
**Pull request by rodiazet**

### CODE REVIEW

File: libsolidity/interface/CompilerStack.cpp [hunk 1]
```diff
@@ -1413,7 +1421,7 @@ void CompilerStack::compileContract(
 
 	Contract& compiledContract = m_contracts.at(_contract.fullyQualifiedName());
 
-	std::shared_ptr<Compiler> compiler = std::make_shared<Compiler>(m_evmVersion, m_revertStrings, m_optimiserSettings);
+	std::shared_ptr<Compiler> compiler = std::make_shared<Compiler>(m_evmVersion, std::nullopt, m_revertStrings, m_optimiserSettings);
```
Comment (ekpyron) [inline libsolidity/interface/CompilerStack.cpp]: I actually didn't mean just passing ``std::nullopt`` here, but not adding the eof version option to ``Compiler`` (and ``CompilerContext``) in the first place (and just have them pass ``std::nullopt`` to the ``Assembly``s they need and such)
And an assert
```
solUnimplementedAssert(!m_eofVersion, "EOF is not supported for legacy code generation");
```
here also wouldn't hurt, together with the ``!m_viaIR`` assert below maybe.

File: libyul/backends/evm/AbstractAssembly.h [hunk 2]
```diff
@@ -99,7 +99,7 @@ class AbstractAssembly
 	/// Append the assembled size as a constant.
 	virtual void appendAssemblySize() = 0;
 	/// Creates a new sub-assembly, which can be referenced using dataSize and dataOffset.
-	virtual std::pair<std::shared_ptr<AbstractAssembly>, SubID> createSubAssembly(bool _creation, std::string _name = "") = 0;
+	virtual std::pair<std::shared_ptr<AbstractAssembly>, SubID> createSubAssembly(bool _creation, std::optional<uint8_t> _eofVersion, std::string _name = "") = 0;
```
Comment (ekpyron) [inline libyul/backends/evm/AbstractAssembly.h]: Also looking a bit closer here: should a sub-assembly get its own EOF version? I'd imagine that it'd always inherit the EOF version from the parent Assembly. (In the ``EthAssemblyAdapter`` implementation you can fetch it from the wrapped ``Assembly``, in ``NoOutputAssembly``, you'd need to store it.)

Or is the idea that technically there could be EOF version 1 containers containing EOF version 2 containers?

File: libevmasm/Assembly.cpp [hunk 3]
```diff
@@ -862,6 +864,8 @@ LinkerObject const& Assembly::assemble() const
 
 	LinkerObject& ret = m_assembledObject;
 
+	bool const eof = m_eofVersion.has_value();
```
Comment (ekpyron) [inline libevmasm/Assembly.cpp:867]:
```suggestion
	bool const eof = m_eofVersion.has_value();
	solAssert(!eof || m_eofVersion == 1, "Invalid EOF version.");
```

**Review by ekpyron — COMMENTED** [review]
Looking at it again in isolation I do have a few more, but only minor comments - I've not yet seen why this would make that one test failing that CI complains about (I had assumed the ``CommonOptions::get().eofVersion()`` would also be empty for now?).

File: libevmasm/Assembly.cpp (hunk 3 above)
Comment (rodiazet) [reply libevmasm/Assembly.cpp:867]: Done

**Comment by rodiazet** [comment]
> Looking at it again in isolation I do have a few more, but only minor comments - I've not yet seen why this would make that one test failing that CI complains about (I had assumed the `CommonOptions::get().eofVersion()` would also be empty for now?).

Problem is in https://github.com/ipsilon/solidity/blob/eof-prep-1/test/libsolidity/Metadata.cpp#L228 test. It doesn't make sense for now. Even changing first evm version to `prague()` does not help because the codegen produce a code with `PushTag` which is not allowed for eof and is asserts in https://github.com/ipsilon/solidity/blob/eof-prep-1/libevmasm/Assembly.cpp#L965.

The test should be added later and adjusted to new EOF version. I can move it to proper commit (later PR) of the `eof-cleaned` branch. Let me know your opinion.

**Comment by ekpyron** [comment]
> > Looking at it again in isolation I do have a few more, but only minor comments - I've not yet seen why this would make that one test failing that CI complains about (I had assumed the `CommonOptions::get().eofVersion()` would also be empty for now?).
> 
> Problem is in https://github.com/ipsilon/solidity/blob/eof-prep-1/test/libsolidity/Metadata.cpp#L228 test. It doesn't make sense for now. Even changing first evm version to `prague()` does not help because the codegen produce a code with `PushTag` which is not allowed for eof and is asserts in https://github.com/ipsilon/solidity/blob/eof-prep-1/libevmasm/Assembly.cpp#L965.
> 
> The test should be added later and adjusted to new EOF version. I can move it to proper commit (later PR) of the `eof-cleaned` branch. Let me know your opinion.

Makes sense, yeah - we need clean CI to merge this, but we can just remove the test for now, merge this PR and bring the test back later, yes!

**Review by ekpyron — APPROVED** [review]

**Pull request merged**